README
These are the instructions for building, installing, and running the OpenPGP
Public Key Server.

Choose a location to install the system.  This should be on the local
disk.  If you plan on running with a full key database, you should
have at least 5 gigabytes free.  If you are planning on running the mail
server, you will need to install the system as root.  Otherwise, you can
install it as any user.  For the rest of these instructions, PREFIX
represents the installation path.


** To Build the OpenPGP Public Key Server:

In the top level, do

	./configure --prefix=PREFIX
	make

On some systems, you may have to use GNU make.  If you see
dependencies failing, this is the problem.

** To Install the OpenPGP Public Key Server:

In the top level, do

	make install

You can look in PREFIX/man for more detailed documentation on the
various programs.  An overview is in pks-intro(8).

As a part of the installation, a template configuration file was
installed in PREFIX/etc/pksd.conf.  This file is fully documented in
pksd.conf(5).  If you don't want to read the manual, the template file
contains comments which describe what you need to change to get your
server to run.

The key server uses syslog for logging.  It logs using the local2
facility (if available), and various priority levels as described in pksd(8).
If you want to get any logging, you should add an appropriate entry to
/etc/syslog.conf on your machine.

If you have configured the mail server component of the key server,
there is one more step you need to take so that the server can
actually get mail.  In your aliases file, create the following lines:

    pgp-public-keys: "|PREFIX/bin/pks-mail.sh PREFIX/etc/pksd.conf"
    pgp: pgp-public-keys

The aliases file is commonly /etc/mail/aliases, /etc/aliases, etc.

You might also need to add aliases for the mail addresses you
configured in pksd.conf.  You will also need to make sure the
permissions on PREFIX/var/incoming allow both the mailer and whatever
user the pksd runs as to insert and delete files.

If you have configured the web server component of the package, you
will want to install the pks commands web page somewhere.  The key
server does not serve arbitrary HTML files, so you must use another
web server for this.  An example HTML file can be found in
pks-commands.html in the distribution.

** To Run the OpenPGP public key server:

Before running the server, read and familiarize yourself with the
information in the DATABASE ADMINISTRATION section of pks-intro(8)
(PREFIX/man/man8/pks-intro.8), in particular the section on
checkpointing.

To create an empty database, run:
PREFIX/bin/pksclient PREFIX/var/db create

If a database already exists, this will remove all its
contents.

To initialize the database with the contents of a keyring, run:
PREFIX/bin/pksclient PREFIX/var/db add keyring.pgp

You can do this multiple times to add the contents of several keyrings.
Because of the transaction semantics, adding very large keyrings will be
inefficient.  You should use the PREFIX/bin/pgpsplit program to split
any very large keyrings into chunks (about 10 MB is fine) and add them
individually.  Adding a 20,000 element keyring on a SparcStation 5
running Solaris takes about an hour.  Your system may take more or
less time.  In the current implementation, this may also be a
memory-intensive task.

To have the daemon run when the system boots, you should run the
following programs in your /etc/rc.local or equivalent file:

PREFIX/bin/pksd PREFIX/etc/pksd.conf
sleep 5
PREFIX/bin/pks-queue-run.sh PREFIX/etc/pksd.conf

You can always run this command by hand if you want.

If you want to use a a SystemV-style init script, pks.init is provided in the
source. 

** For more information:

There is an overview of the system in pks-intro(8).

** If you have any problems, questions, or comments:

Send an e-mail to pgp-keyserver-folk@flame.org.  This is a public listserv
dedicated to all keyserver issues, regardless of keyserver software. As such,
normal listserv etiquette is expected. Be sure to mention that you're running
PKS and provide the version number.

** Acknowledgements

Thanks to Michael Helm <helm@fionn.es.net>, Marc Dyksterhouse
<marcd@pgp.com>, "Teun Nijssen" <Teun.Nijssen@kub.nl>, George Ross
<gdmr@dcs.ed.ac.uk>, Marcel Waldvogel <mwa@tik.ee.ethz.ch>, Olaf
Dabrunz <dabrunz@santana.rrz.uni-hamburg.de>, Paul Leyland
<pcl@sable.ox.ac.uk>, Hironobu Suzuki
<hironobu@h2np.suginami.tokyo.jp>, Josef Pojsl
<josef.pojsl@skynet.cz>, and all the other keyserver testers and
maintainers for their help testing the keyserver, reporting bugs, and
especially contributing fixes and new features.

Contributors since version 0.9.4 are listed with their contributions in the
ChangeLog file.

================================================================

$Id: README,v 1.3 2003/02/04 22:52:42 rlaager Exp $

Copyright (c) 1996, 1997, 1998, 1999, Marc Horowitz.  All rights reserved.
See the LICENSE file in the release for redistribution information.


NEWS
For general information about the keyserver, visit
<http://sourceforge.net/projects/pks/>.

***** Changes in version 0.9.6

- A new machine-readable format exists for easy parsing by GPG, et al.
- Lots of bug squashing, documentation updates, etc.
- Temporary fixes for corruption that occurs with subkeys. (The packets
  are dropped instead of being added in a corrupt format.)
  READ THE FILE "UPGRADING" FOR INFORMATION ON UPGRADING TO 0.9.6!!!

***** Changes in version 0.9.5

- Allow 16-digit and 40-digit keyids/fingerprints.  Currently only the
  low 8 digits are used, but this should smooth a transition to a
  future pksd that allows long keyid or fingerprint searches.
- Nickolai Zeldovich's fix for garbage mixed in with large keys.
- Add missing transaction hooks (patch2)
- Make checkpoint code retry if necessary (patch2)
- Change the way memory is allocated in some places to reduce the
number of extra calls to malloc and free (patch2)
- Fix a bug in db 2.7.5 which causes db_recover to fail under some
circumstances. (patch2)
- Stopped flooding caused by multiple revokation certs (flood.patch)
- In kd_search, fixed bug in using maxid when keyid not found
(kdsearch_error.patch)
- WWW key submissions disabled by default (JHPatch1)
- nicer formatting of SHA-1 fingerprints (JHPatch1)
- pksdctl() usage page improvements (JHPatch1)
- manual page improvements (JHPatch1)
- Fixed buffer overflow from long arguments to search in HDP requests,
as mentioned on bugtraq (kd_search by Jason Harris)
- Fixed problem of randomly not sending updated sigs to syncsites due
to improperly initialized memory. (patch_pf20020615)
- Added compile-time support for X509 recognition (x509.patch)
- Fixed makefiles to more support rpm build roots (pks-bldroot.patch)
- Changed install to put db2 in own subdir (pks-db2-install.patch)
- Added libwrap support (pks-libwrap.patch)
- Added sample init script (pks-rc.patch)
- Beautified some of the web output (pks-beauty.patch)
- Fixed a buffer overflow (from Teun Nijssen's source tree)
- Client connection aborts are now handled properly on newer versions
of Solaris (from Teun Nijssen's source tree - original change by
Xander Jansen of SURFnet)
- Added checkpointing after so many transactions. (patch3)
- Added more logging to kvcv.c (patch3)
- Added some messages regarding the non-support of X.509. (patch3)
- Added max_last_reply_keys configuration option (patch3)
- The spec file was radically updated for the upcoming pks-0.9.5rc1.
  o The database is now stored in /var/lib/pks instead of /home/keyserver
  o The PREFIX is now /usr instead of /usr/local
  o For more details, see the %changelog that is within the spec file.

***** Changes in version 0.9.4

- Converted to Sleepycat Software Berkeley DB version 2.7.5.
- mmap()d input files instead of reading them when possible.
- The database may be split into multiple smaller files.
- When a key does not parse, return a soft error to the user and
continue processing the input keyblock.
- When extending lists in the database, do so in chunks, rather than
one entry at a time.  This improves performance when adding keys.
- Allow the administrator to limit the maximum duration of the LAST
mail server command.
- Allow the administrator to limit the number of keys returned by
index, verbose index, get, and last requests.
- Use the log, lock, and transaction features of Berkeley DB.

***** Changes in version 0.9.2

- Added support for pgp 5's DSA and ElGamal keys.
- Added support to short-circuit the normal output paths in
pksclient, and instead output indexes and binary keys directly to
stdout.  This saves a lot of virtual memory.
- Removed many gmake-isms from the Makefile.
- In the index pages from the web server, links in the User ID look
up by KeyID instead of name, to eliminate ambiguity.
- Fixed defunct process leak on some platforms, notably solaris.
- Check for and, if necessary, use bsd4.2/ultrix 2-argument openlog.
- The help files now must have a (possibly empty) MIME header block.
This allows help files to use non-ascii character sets.
- keyrings returned by the mail server have content-type
application/pgp-keys, as per rfc2015.
- fixed a double-memory-free.
- pksd is more careful about shutting down cleanly when it receives a
non-fatal signal (INT, QUIT, HUP, TERM).


md5c.c
const char rcsid_md5c_c[] = "$Id: md5c.c,v 1.1.1.1 2002/09/04 20:48:53 dtype Exp $";

/* 
 * Copyright (c) 1996, Marc Horowitz.  All rights reserved.
 * See the LICENSE file in the release for redistribution information.
 */


/* MD5C.C - RSA Data Security, Inc., MD5 message-digest algorithm
 */

/* Copyright (C) 1991-2, RSA Data Security, Inc. Created 1991. All
rights reserved.

License to copy and use this software is granted provided that it
is identified as the "RSA Data Security, Inc. MD5 Message-Digest
Algorithm" in all material mentioning or referencing this software
or this function.

License is also granted to make and use derivative works provided
that such works are identified as "derived from the RSA Data
Security, Inc. MD5 Message-Digest Algorithm" in all material
mentioning or referencing the derived work.

RSA Data Security, Inc. makes no representations concerning either
the merchantability of this software or the suitability of this
software for any particular purpose. It is provided "as is"
without express or implied warranty of any kind.

These notices must be retained in any copies of any part of this
documentation and/or software.
 */

#include <string.h>

#include "md5.h"

#define PROTO_LIST(list) list

/* Constants for MD5Transform routine.
 */

#define S11 7
#define S12 12
#define S13 17
#define S14 22
#define S21 5
#define S22 9
#define S23 14
#define S24 20
#define S31 4
#define S32 11
#define S33 16
#define S34 23
#define S41 6
#define S42 10
#define S43 15
#define S44 21

static void MD5Transform PROTO_LIST ((UINT4 [4], unsigned char [64]));
static void Encode PROTO_LIST
  ((unsigned char *, UINT4 *, unsigned int));
static void Decode PROTO_LIST
  ((UINT4 *, unsigned char *, unsigned int));
#if 0
static void MD5_memcpy PROTO_LIST ((POINTER, POINTER, unsigned int));
static void MD5_memset PROTO_LIST ((POINTER, int, unsigned int));
#else
#define MD5_memcpy memcpy
#define MD5_memset memset
#endif

static unsigned char PADDING[64] = {
  0x80, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
  0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
};

/* F, G, H and I are basic MD5 functions.
 */
#define F(x, y, z) (((x) & (y)) | ((~x) & (z)))
#define G(x, y, z) (((x) & (z)) | ((y) & (~z)))
#define H(x, y, z) ((x) ^ (y) ^ (z))
#define I(x, y, z) ((y) ^ ((x) | (~z)))

/* ROTATE_LEFT rotates x left n bits.
 */
#define ROTATE_LEFT(x, n) (((x) << (n)) | ((x) >> (32-(n))))

/* FF, GG, HH, and II transformations for rounds 1, 2, 3, and 4.
Rotation is separate from addition to prevent recomputation.
 */
#define FF(a, b, c, d, x, s, ac) { \
 (a) += F ((b), (c), (d)) + (x) + (UINT4)(ac); \
 (a) = ROTATE_LEFT ((a), (s)); \
 (a) += (b); \
  }
#define GG(a, b, c, d, x, s, ac) { \
 (a) += G ((b), (c), (d)) + (x) + (UINT4)(ac); \
 (a) = ROTATE_LEFT ((a), (s)); \
 (a) += (b); \
  }
#define HH(a, b, c, d, x, s, ac) { \
 (a) += H ((b), (c), (d)) + (x) + (UINT4)(ac); \
 (a) = ROTATE_LEFT ((a), (s)); \
 (a) += (b); \
  }
#define II(a, b, c, d, x, s, ac) { \
 (a) += I ((b), (c), (d)) + (x) + (UINT4)(ac); \
 (a) = ROTATE_LEFT ((a), (s)); \
 (a) += (b); \
  }

/* MD5 initialization. Begins an MD5 operation, writing a new context.
 */
void MD5Init (context)
MD5_CTX *context;                                        /* context */
{
  context->count[0] = context->count[1] = 0;
  /* Load magic initialization constants.
*/
  context->state[0] = 0x67452301;
  context->state[1] = 0xefcdab89;
  context->state[2] = 0x98badcfe;
  context->state[3] = 0x10325476;
}

/* MD5 block update operation. Continues an MD5 message-digest
  operation, processing another message block, and updating the
  context.
 */
void MD5Update (context, input, inputLen)
MD5_CTX *context;                                        /* context */
unsigned char *input;                                /* input block */
unsigned int inputLen;                     /* length of input block */
{
  unsigned int i, index, partLen;

  /* Compute number of bytes mod 64 */
  index = (unsigned int)((context->count[0] >> 3) & 0x3F);

  /* Update number of bits */
  if ((context->count[0] += ((UINT4)inputLen << 3))
   < ((UINT4)inputLen << 3))
 context->count[1]++;
  context->count[1] += ((UINT4)inputLen >> 29);

  partLen = 64 - index;

  /* Transform as many times as possible.
*/
  if (inputLen >= partLen) {
 MD5_memcpy
   ((POINTER)&context->buffer[index], (POINTER)input, partLen);
 MD5Transform (context->state, context->buffer);

 for (i = partLen; i + 63 < inputLen; i += 64)
   MD5Transform (context->state, &input[i]);

 index = 0;
  }
  else
 i = 0;

  /* Buffer remaining input */
  MD5_memcpy
 ((POINTER)&context->buffer[index], (POINTER)&input[i],
  inputLen-i);
}

/* MD5 finalization. Ends an MD5 message-digest operation, writing the
  the message digest and zeroizing the context.
 */
void MD5Final (digest, context)
unsigned char digest[16];                         /* message digest */
MD5_CTX *context;                                       /* context */
{
  unsigned char bits[8];
  unsigned int index, padLen;

  /* Save number of bits */
  Encode (bits, context->count, 8);

  /* Pad out to 56 mod 64.
*/
  index = (unsigned int)((context->count[0] >> 3) & 0x3f);
  padLen = (index < 56) ? (56 - index) : (120 - index);
  MD5Update (context, PADDING, padLen);

  /* Append length (before padding) */
  MD5Update (context, bits, 8);
  /* Store state in digest */
  Encode (digest, context->state, 16);

  /* Zeroize sensitive information.
*/
  MD5_memset ((POINTER)context, 0, sizeof (*context));
}

/* MD5 basic transformation. Transforms state based on block.
 */
static void MD5Transform (state, block)
UINT4 state[4];
unsigned char block[64];
{
  UINT4 a = state[0], b = state[1], c = state[2], d = state[3], x[16];

  Decode (x, block, 64);

  /* Round 1 */
  FF (a, b, c, d, x[ 0], S11, 0xd76aa478); /* 1 */
  FF (d, a, b, c, x[ 1], S12, 0xe8c7b756); /* 2 */
  FF (c, d, a, b, x[ 2], S13, 0x242070db); /* 3 */
  FF (b, c, d, a, x[ 3], S14, 0xc1bdceee); /* 4 */
  FF (a, b, c, d, x[ 4], S11, 0xf57c0faf); /* 5 */
  FF (d, a, b, c, x[ 5], S12, 0x4787c62a); /* 6 */
  FF (c, d, a, b, x[ 6], S13, 0xa8304613); /* 7 */
  FF (b, c, d, a, x[ 7], S14, 0xfd469501); /* 8 */
  FF (a, b, c, d, x[ 8], S11, 0x698098d8); /* 9 */
  FF (d, a, b, c, x[ 9], S12, 0x8b44f7af); /* 10 */
  FF (c, d, a, b, x[10], S13, 0xffff5bb1); /* 11 */
  FF (b, c, d, a, x[11], S14, 0x895cd7be); /* 12 */
  FF (a, b, c, d, x[12], S11, 0x6b901122); /* 13 */
  FF (d, a, b, c, x[13], S12, 0xfd987193); /* 14 */
  FF (c, d, a, b, x[14], S13, 0xa679438e); /* 15 */
  FF (b, c, d, a, x[15], S14, 0x49b40821); /* 16 */

 /* Round 2 */
  GG (a, b, c, d, x[ 1], S21, 0xf61e2562); /* 17 */
  GG (d, a, b, c, x[ 6], S22, 0xc040b340); /* 18 */
  GG (c, d, a, b, x[11], S23, 0x265e5a51); /* 19 */
  GG (b, c, d, a, x[ 0], S24, 0xe9b6c7aa); /* 20 */
  GG (a, b, c, d, x[ 5], S21, 0xd62f105d); /* 21 */
  GG (d, a, b, c, x[10], S22,  0x2441453); /* 22 */
  GG (c, d, a, b, x[15], S23, 0xd8a1e681); /* 23 */
  GG (b, c, d, a, x[ 4], S24, 0xe7d3fbc8); /* 24 */
  GG (a, b, c, d, x[ 9], S21, 0x21e1cde6); /* 25 */
  GG (d, a, b, c, x[14], S22, 0xc33707d6); /* 26 */
  GG (c, d, a, b, x[ 3], S23, 0xf4d50d87); /* 27 */
  GG (b, c, d, a, x[ 8], S24, 0x455a14ed); /* 28 */
  GG (a, b, c, d, x[13], S21, 0xa9e3e905); /* 29 */
  GG (d, a, b, c, x[ 2], S22, 0xfcefa3f8); /* 30 */
  GG (c, d, a, b, x[ 7], S23, 0x676f02d9); /* 31 */
  GG (b, c, d, a, x[12], S24, 0x8d2a4c8a); /* 32 */

  /* Round 3 */
  HH (a, b, c, d, x[ 5], S31, 0xfffa3942); /* 33 */
  HH (d, a, b, c, x[ 8], S32, 0x8771f681); /* 34 */
  HH (c, d, a, b, x[11], S33, 0x6d9d6122); /* 35 */
  HH (b, c, d, a, x[14], S34, 0xfde5380c); /* 36 */
  HH (a, b, c, d, x[ 1], S31, 0xa4beea44); /* 37 */
  HH (d, a, b, c, x[ 4], S32, 0x4bdecfa9); /* 38 */
  HH (c, d, a, b, x[ 7], S33, 0xf6bb4b60); /* 39 */
  HH (b, c, d, a, x[10], S34, 0xbebfbc70); /* 40 */
  HH (a, b, c, d, x[13], S31, 0x289b7ec6); /* 41 */
  HH (d, a, b, c, x[ 0], S32, 0xeaa127fa); /* 42 */
  HH (c, d, a, b, x[ 3], S33, 0xd4ef3085); /* 43 */
  HH (b, c, d, a, x[ 6], S34,  0x4881d05); /* 44 */
  HH (a, b, c, d, x[ 9], S31, 0xd9d4d039); /* 45 */
  HH (d, a, b, c, x[12], S32, 0xe6db99e5); /* 46 */
  HH (c, d, a, b, x[15], S33, 0x1fa27cf8); /* 47 */
  HH (b, c, d, a, x[ 2], S34, 0xc4ac5665); /* 48 */

  /* Round 4 */
  II (a, b, c, d, x[ 0], S41, 0xf4292244); /* 49 */
  II (d, a, b, c, x[ 7], S42, 0x432aff97); /* 50 */
  II (c, d, a, b, x[14], S43, 0xab9423a7); /* 51 */
  II (b, c, d, a, x[ 5], S44, 0xfc93a039); /* 52 */
  II (a, b, c, d, x[12], S41, 0x655b59c3); /* 53 */
  II (d, a, b, c, x[ 3], S42, 0x8f0ccc92); /* 54 */
  II (c, d, a, b, x[10], S43, 0xffeff47d); /* 55 */
  II (b, c, d, a, x[ 1], S44, 0x85845dd1); /* 56 */
  II (a, b, c, d, x[ 8], S41, 0x6fa87e4f); /* 57 */
  II (d, a, b, c, x[15], S42, 0xfe2ce6e0); /* 58 */
  II (c, d, a, b, x[ 6], S43, 0xa3014314); /* 59 */
  II (b, c, d, a, x[13], S44, 0x4e0811a1); /* 60 */
  II (a, b, c, d, x[ 4], S41, 0xf7537e82); /* 61 */
  II (d, a, b, c, x[11], S42, 0xbd3af235); /* 62 */
  II (c, d, a, b, x[ 2], S43, 0x2ad7d2bb); /* 63 */
  II (b, c, d, a, x[ 9], S44, 0xeb86d391); /* 64 */

  state[0] += a;
  state[1] += b;
  state[2] += c;
  state[3] += d;

  /* Zeroize sensitive information.
*/
  MD5_memset ((POINTER)x, 0, sizeof (x));
}

/* Encodes input (UINT4) into output (unsigned char). Assumes len is
  a multiple of 4.
 */
static void Encode (output, input, len)
unsigned char *output;
UINT4 *input;
unsigned int len;
{
  unsigned int i, j;

  for (i = 0, j = 0; j < len; i++, j += 4) {
 output[j] = (unsigned char)(input[i] & 0xff);
 output[j+1] = (unsigned char)((input[i] >> 8) & 0xff);
 output[j+2] = (unsigned char)((input[i] >> 16) & 0xff);
 output[j+3] = (unsigned char)((input[i] >> 24) & 0xff);
  }
}

/* Decodes input (unsigned char) into output (UINT4). Assumes len is
  a multiple of 4.
 */
static void Decode (output, input, len)
UINT4 *output;
unsigned char *input;
unsigned int len;
{
  unsigned int i, j;

  for (i = 0, j = 0; j < len; i++, j += 4)
 output[i] = ((UINT4)input[j]) | (((UINT4)input[j+1]) << 8) |
   (((UINT4)input[j+2]) << 16) | (((UINT4)input[j+3]) << 24);
}

#if 0

/* Note: Replace "for loop" with standard memcpy if possible.
 */

static void MD5_memcpy (output, input, len)
POINTER output;
POINTER input;
unsigned int len;
{
  unsigned int i;

  for (i = 0; i < len; i++)
     output[i] = input[i];
}

/* Note: Replace "for loop" with standard memset if possible.
 */
static void MD5_memset (output, value, len)
POINTER output;
int value;
unsigned int len;
{
  unsigned int i;

  for (i = 0; i < len; i++)
 ((char *)output)[i] = (char)value;
}

#endif
